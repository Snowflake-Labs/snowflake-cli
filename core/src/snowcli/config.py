from __future__ import annotations

import os
from pathlib import Path
from typing import Optional, Any, Dict, Union, List

import tomlkit
from snowflake.connector.errors import MissingConfigOptionError
from tomlkit import dump, table, TOMLDocument
from tomlkit.exceptions import NonExistentKey
from tomlkit.items import Table
from tomlkit.container import Container
import logging

from snowcli.exception import MissingConfiguration, UnsupportedConfigSectionTypeError
from snowflake.connector.constants import CONFIG_FILE
from snowflake.connector.config_manager import ConfigManager

log = logging.getLogger(__name__)


class CliConfigManager(ConfigManager):
    def __init__(self, file_path: Path = CONFIG_FILE):
        super().__init__(name="SNOWCLI_PARSER", file_path=file_path)
        self._add_options()

    def from_context(self, config_path_override: Optional[Path]):
        if config_path_override:
            self.file_path = config_path_override
        if not self.file_path.exists():
            self._initialise_config()
        self.read_config()

    def get_section(self, *path) -> dict:
        section = self._find_section(*path)
        if type(section) is Container:
            return {
                s: self._merge_section_with_env(section[s], *path, s) for s in section
            }
        elif type(section) is Table:
            return self._merge_section_with_env(section, *path)
        raise UnsupportedConfigSectionTypeError(type(section))

    def section_exists(self, *path) -> bool:
        try:
            self._find_section(*path)
            return True
        except (NonExistentKey, MissingConfigOptionError):
            return False

    def get(self, *path, key: str, default: Optional[Any] = None) -> Any:
        """Looks for given key under nested path in toml file."""
        env_variable = self._get_env_value(*path, key=key)
        if env_variable:
            return env_variable
        try:
            return self.get_section(*path)[key]
        except (NonExistentKey, MissingConfigOptionError):
            if default:
                return default
            raise

    def get_connection(self, connection_name: str) -> dict:
        try:
            return self.get_section("connections", connection_name)
        except NonExistentKey:
            raise MissingConfiguration(
                f"Connection {connection_name} is not configured"
            )

    def add_connection(self, name: str, parameters: dict):
        if not self.section_exists("connections"):
            self._initialize_connection_section()
        self._find_section("connections").add(name, parameters)
        self._dump_config()

    # TODO: think about extracting plugin specific methods to another layer
    def enable_plugin(self, name: str):
        enabled_plugins = self.get_enabled_plugins()
        if name not in enabled_plugins:
            enabled_plugins.append(name)
        self._replace_enabled_plugins(enabled_plugins)

    def disable_plugin(self, name: str):
        enabled_plugins = self.get_enabled_plugins()
        if name in enabled_plugins:
            enabled_plugins.remove(name)
        self._replace_enabled_plugins(enabled_plugins)

    def _replace_enabled_plugins(self, new_enabled_plugins: List[str]):
        plugins_section = self._find_section("plugins")
        plugins_section.remove("enabled")
        plugins_section.add("enabled", new_enabled_plugins)
        self._dump_config()

    def get_enabled_plugins(self) -> List[str]:
        if not self.section_exists("plugins"):
            self._initialize_plugins_section()
        plugins_section = self.get_section("plugins")
        if "enabled" in plugins_section:
            return plugins_section["enabled"]
        else:
            return []

    def _add_options(self):
        self.add_option(
            name="options",
            parse_str=tomlkit.parse,
        )
        self.add_option(
            name="plugins",
            parse_str=tomlkit.parse,
        )
        self.add_option(
            name="connections",
            parse_str=tomlkit.parse,
        )

    def _find_section(self, *path) -> TOMLDocument:
        section = self
        idx = 0
        while idx < len(path):
            section = section[path[idx]]
            idx += 1
        return section

    def _merge_section_with_env(
        self, section: Union[Table, Any], *path
    ) -> Dict[str, str]:
        if isinstance(section, Table):
            env_variables = self._get_envs_for_path(*path)
            section_copy = section.copy()
            section_copy.update(env_variables)
            return section_copy
        # It's a atomic value
        return section

    def _get_env_value(self, *path, key: str):
        env_variable_name = (
            "SNOWFLAKE_" + "_".join(p.upper() for p in path) + f"_{key.upper()}"
        )
        return os.environ.get(env_variable_name)

    def _get_envs_for_path(self, *path) -> dict:
        env_variables_prefix = "SNOWFLAKE_" + "_".join(p.upper() for p in path)
        return {
            k.replace(f"{env_variables_prefix}_", "").lower(): os.environ[k]
            for k in os.environ.keys()
            if k.startswith(env_variables_prefix)
        }

    def _initialize_connection_section(self):
        self.conf_file_cache = TOMLDocument()
        self.conf_file_cache.append("connections", table())

    def _initialize_plugins_section(self):
        self.conf_file_cache.append("plugins", table())

    def _initialise_config(self):
        os.makedirs(os.path.dirname(self.file_path), exist_ok=True)
        self._initialize_connection_section()
        self._initialize_plugins_section()
        self._dump_config()
        log.info(f"Created Snowflake configuration file at {cli_config.file_path}")

    def _dump_config(self):
        with open(self.file_path, "w+") as fh:
            dump(self.conf_file_cache, fh)


def config_init(config_file: Path):
    """
    Initializes the app configuration. Config provided via cli flag takes precedence.
    If config file does not exist we create an empty one.
    """
    cli_config.from_context(config_path_override=config_file)


cli_config: CliConfigManager = CliConfigManager()  # type: ignore

_DEFAULT_CONNECTION = "dev"


def get_default_connection():
    return cli_config.get(
        "options", key="default_connection", default=_DEFAULT_CONNECTION
    )
