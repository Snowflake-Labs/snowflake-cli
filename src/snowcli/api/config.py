from __future__ import annotations

import logging
import os
from pathlib import Path
from typing import Any, Dict, Optional, Union

import tomlkit
from snowcli.api.exceptions import (
    MissingConfiguration,
    UnsupportedConfigSectionTypeError,
)
from snowflake.connector.config_manager import CONFIG_MANAGER
from snowflake.connector.errors import MissingConfigOptionError
from tomlkit import TOMLDocument, dump
from tomlkit.container import Container
from tomlkit.exceptions import NonExistentKey
from tomlkit.items import Table

log = logging.getLogger(__name__)


class Empty:
    pass


CONFIG_MANAGER.add_option(
    name="snowcli",
    parse_str=tomlkit.parse,
    default=dict(),
)


def config_init(config_file: Path):
    """
    Initializes the app configuration. Config provided via cli flag takes precedence.
    If config file does not exist we create an empty one.
    """
    if not config_file:
        return

    CONFIG_MANAGER.file_path = config_file
    if not config_file.exists():
        _initialise_config(config_file)
    CONFIG_MANAGER.read_config()


def add_connection(name: str, parameters: dict):
    conf_file_cache = CONFIG_MANAGER.conf_file_cache
    if conf_file_cache.get("connections") is None:
        conf_file_cache["connections"] = {}
    conf_file_cache["connections"][name] = parameters
    _dump_config(conf_file_cache)


def connection_exists(connection_name: str) -> bool:
    return config_section_exists("connections", connection_name)


def config_section_exists(*path) -> bool:
    try:
        _find_section(*path)
        return True
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        return False


def get_connection(connection_name: str) -> dict:
    try:
        return get_config_section("connections", connection_name)
    except KeyError:
        raise MissingConfiguration(f"Connection {connection_name} is not configured")


def get_config_section(*path) -> dict:
    section = _find_section(*path)
    if type(section) is Container:
        return {s: _merge_section_with_env(section[s], *path, s) for s in section}
    elif type(section) is Table:
        return _merge_section_with_env(section, *path)
    raise UnsupportedConfigSectionTypeError(type(section))


def get_config_value(*path, key: str, default: Optional[Any] = Empty) -> Any:
    """Looks for given key under nested path in toml file."""
    env_variable = _get_env_value(*path, key=key)
    if env_variable:
        return env_variable
    try:
        return get_config_section(*path)[key]
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        if default is not Empty:
            return default
        raise


def _initialise_config(config_file: Path) -> None:
    os.makedirs(os.path.dirname(config_file), exist_ok=True)
    config_file.touch()
    log.info("Created Snowflake configuration file at %s", CONFIG_MANAGER.file_path)


def _get_env_value(*path, key: str) -> str | None:
    env_variable_name = (
        "SNOWFLAKE_" + "_".join(p.upper() for p in path) + f"_{key.upper()}"
    )
    return os.environ.get(env_variable_name)


def _find_section(*path) -> TOMLDocument:
    section = CONFIG_MANAGER
    idx = 0
    while idx < len(path):
        section = section[path[idx]]
        idx += 1
    return section


def _merge_section_with_env(section: Union[Table, Any], *path) -> Dict[str, str]:
    if isinstance(section, Table):
        env_variables = _get_envs_for_path(*path)
        section_copy = section.copy()
        section_copy.update(env_variables)
        return section_copy.unwrap()
    # It's a atomic value
    return section


def _get_envs_for_path(*path) -> dict:
    env_variables_prefix = "SNOWFLAKE_" + "_".join(p.upper() for p in path)
    return {
        k.replace(f"{env_variables_prefix}_", "").lower(): os.environ[k]
        for k in os.environ.keys()
        if k.startswith(env_variables_prefix)
    }


def _dump_config(conf_file_cache: Dict):
    with open(CONFIG_MANAGER.file_path, "w+") as fh:
        dump(conf_file_cache, fh)
