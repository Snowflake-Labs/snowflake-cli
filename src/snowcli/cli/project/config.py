from .util import clean_identifier, get_env_username
from pathlib import Path
from typing import List
from strictyaml import (
    YAML,
    load,
    as_document,
)

from .schema import (
    project_schema,
    project_override_schema,
)


DEFAULT_USERNAME = "unknown_user"


def merge_left(target: dict | YAML, source: dict | YAML) -> None:
    """
    Recursively merges key/value pairs from source into target.
    Modifies the original dict-like "target".
    """
    for k, v in source.items():
        if k in target and (isinstance(v, dict) or isinstance(v, YAML)):
            # assumption: all inputs have been validated.
            assert isinstance(target[k], dict) or isinstance(target[k], YAML)
            merge_left(target[k], v)
        else:
            target[k] = v


def load_project_config(paths: List[Path]) -> dict:
    """
    Loads a project config, optionally overriding values. Configuration is merged
    in order of left to right (increasing precedence).
    """
    if len(paths) == 0:
        raise ValueError("Need at least one configuration file.")

    with open(paths[0], "r") as base_yml:
        config = load(base_yml.read(), project_schema)

    for override_path in paths[1:]:
        with open(override_path, "r") as override_yml:
            overrides = load(override_yml.read(), project_override_schema)
            merge_left(config, overrides)

        # TODO: how to show good error messages here?
        config.revalidate(project_schema)

    return config.data


def generate_local_override_yml(project: dict | YAML, conn: dict) -> YAML:
    user = clean_identifier(get_env_username() or DEFAULT_USERNAME)
    role = conn.get("role", "accountadmin")  # TODO: actual connection

    local: dict = {}
    if "native_app" in project:
        name = clean_identifier(project["native_app"]["name"])
        local["native_app"] = {
            "application": {
                "name": f"{name}_{user}",
                "role": role,
                "debug": True,
                # TODO: warehouse from actual connection
            },
            "package": {"name": f"{name}_pkg_{user}", "role": role},
        }

    return as_document(local, project_override_schema)
