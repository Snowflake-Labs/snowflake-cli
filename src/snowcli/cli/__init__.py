from __future__ import annotations

import logging
import pkgutil
import sys
import typer
from collections.abc import Container
from pathlib import Path

from snowcli import __about__
from snowcli.cli import (
    connection,
    render,
    sql,
    streamlit,
)
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.common.snow_cli_global_context import (
    snow_cli_global_context_manager,
    SnowCliGlobalContext,
)
from snowcli.cli.main.snow_cli_main_typer import SnowCliMainTyper
from snowcli.cli import loggers
from snowcli.cli.snowpark import app as snowpark_app
from snowcli.config import AppConfig
from snowcli.output.formats import OutputFormat
from snowcli.connection_config import ConnectionConfigs

app = SnowCliMainTyper()
log = logging.getLogger(__name__)


def _version_callback(value: bool):
    if value:
        typer.echo(f"SnowCLI Version: {__about__.VERSION}")
        raise typer.Exit()


def setup_global_context(debug: bool):
    def modifications(context: SnowCliGlobalContext) -> SnowCliGlobalContext:
        context.enable_tracebacks = debug
        return context

    snow_cli_global_context_manager.update_global_context(modifications)


@app.command()
def login(
    snowsql_config_path: Path = typer.Option(
        "~/.snowsql/config",
        "--config",
        "-c",
        prompt="Path to Snowsql config",
        help="snowsql config file",
    ),
    connection_name: str = typer.Option(
        ...,
        "--connection",
        "-C",
        prompt="Connection name (for entry in snowsql config)",
        help="connection name from snowsql config file",
    ),
):
    """
    Select a Snowflake connection to use with SnowCLI.
    """
    if not snowsql_config_path.expanduser().exists():
        log.error(f"Path to snowsql config does not exist: {snowsql_config_path}")
        raise typer.Abort()

    connection_configs = ConnectionConfigs(snowsql_config_path)
    if not connection_configs.connection_exists(connection_name):
        log.error(
            f"Connection not found in {snowsql_config_path}: {connection_name}. "
            "You can add with `snow connection add`."
        )
        raise typer.Abort()

    cfg = AppConfig()
    cfg.config["snowsql_config_path"] = str(snowsql_config_path.expanduser())
    cfg.config["snowsql_connection_name"] = connection_name
    cfg.save()
    log.info(f"Using connection name {connection_name} in {snowsql_config_path}")
    log.info(f"Wrote {cfg.path}")


@app.command()
def configure(
    environment: str = typer.Option(
        "dev",
        "-e",
        "--environment",
        help="Name of environment (e.g. dev, prod, staging)",
    ),
    database: str = typer.Option(
        ...,
        "--database",
        prompt="Snowflake database",
        help="Snowflake database",
    ),
    schema: str = typer.Option(
        ...,
        "--schema",
        prompt="Snowflake schema",
        help="Snowflake schema",
    ),
    role: str = typer.Option(
        ...,
        "--role",
        prompt="Snowflake role",
        help="Snowflake role",
    ),
    warehouse: str = typer.Option(
        ...,
        "--warehouse",
        prompt="Snowflake warehouse",
        help="Snowflake warehouse",
    ),
):
    """
    Configure an environment to use with your Snowflake connection.
    """
    log.info(f"Configuring environment {environment}...")
    cfg = AppConfig()
    if environment in cfg.config and not typer.confirm(
        "Environment {environment} already exists. Overwrite?",
    ):
        log.error("Cancelling...")
        raise typer.Abort()

    cfg.config[environment] = {}
    cfg.config[environment]["database"] = database
    cfg.config[environment]["schema"] = schema
    cfg.config[environment]["role"] = role
    cfg.config[environment]["warehouse"] = warehouse
    cfg.save()
    log.info(f"Wrote environment {environment} to {cfg.path}")


@app.callback()
def default(
    version: bool = typer.Option(
        None,
        "--version",
        help="Prints version of the snowcli",
        callback=_version_callback,
        is_eager=True,
    ),
    output_format: OutputFormat = typer.Option(
        OutputFormat.TABLE.value,
        "--format",
        help="Specifies output format",
        case_sensitive=False,
        is_eager=True,
    ),
    configuration_file: Path = typer.Option(
        None,
        "--config-file",
        help="Specifies snowcli configuration file that should be used",
        exists=True,
        dir_okay=False,
        is_eager=True,
    ),
    verbose: bool = typer.Option(
        None,
        "--verbose",
        "-v",
        help="Print logs from level info and higher",
    ),
    debug: bool = typer.Option(
        None,
        "--debug",
        help="Print logs from level debug and higher, logs contains additional information",
    ),
) -> None:
    """
    SnowCLI - A CLI for Snowflake
    """
    loggers.create_loggers(verbose, debug)
    setup_global_context(debug=debug)


MODULE_IGNORE_SET = frozenset(("main", "procedure_coverage"))


def register_cli_typers(ignore_container: Container[str] = MODULE_IGNORE_SET) -> None:
    for _, name, _ in pkgutil.walk_packages(__path__):
        if name not in ignore_container:
            cli_app = __import__(f"{__name__}.{name}", fromlist=["_trash"])
            try:
                app.add_typer(cli_app.app, name=name)
            except AttributeError:
                # Ignore modules that don't define app global
                pass


register_cli_typers()

app.command("sql")(sql.execute_sql)
app.add_typer(snowpark_app)


if __name__ == "__main__":
    app()

if getattr(sys, "frozen", False):
    app(sys.argv[1:])
