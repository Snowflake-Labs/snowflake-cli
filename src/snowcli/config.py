from __future__ import annotations

import logging
import os
from pathlib import Path
from typing import Any, Dict, Optional, Union

import tomlkit
from snowcli.exception import MissingConfiguration, UnsupportedConfigSectionTypeError
from snowflake.connector.config_manager import CONFIG_MANAGER
from snowflake.connector.errors import MissingConfigOptionError
from tomlkit import TOMLDocument, dump
from tomlkit.container import Container
from tomlkit.exceptions import NonExistentKey
from tomlkit.items import Table

log = logging.getLogger(__name__)


class Empty:
    pass


# <<<<<<< HEAD
# class CliConfigManager(ConfigManager):
#     def __init__(self, file_path: Path = CONFIG_FILE):
#         super().__init__(name="SNOWCLI_PARSER", file_path=file_path)
#         self._add_options()
#
#     def from_context(self, config_path_override: Optional[Path]):
#         if config_path_override:
#             self.file_path = config_path_override
#         if not self.file_path.exists():
#             self._initialise_config()
#         self.read_config()
#
#     def get_section(self, *path) -> dict:
#         section = self._find_section(*path)
#         if type(section) is Container:
#             return {
#                 s: self._merge_section_with_env(section[s], *path, s) for s in section
#             }
#         elif type(section) is Table:
#             return self._merge_section_with_env(section, *path)
#         raise UnsupportedConfigSectionTypeError(type(section))
#
#     def section_exists(self, *path) -> bool:
#         try:
#             self._find_section(*path)
#             return True
#         except (NonExistentKey, MissingConfigOptionError):
#             return False
#
#     def get(self, *path, key: str, default: Optional[Any] = Empty) -> Any:
#         """Looks for given key under nested path in toml file."""
#         env_variable = self._get_env_value(*path, key=key)
#         if env_variable:
#             return env_variable
#         try:
#             return self.get_section(*path)[key]
#         except (KeyError, NonExistentKey, MissingConfigOptionError):
#             if default is not Empty:
#                 return default
#             raise
#
#     def get_connection(self, connection_name: str) -> dict:
#         try:
#             return self.get_section("connections", connection_name)
#         except NonExistentKey:
#             raise MissingConfiguration(
#                 f"Connection {connection_name} is not configured"
#             )
#
#     def add_connection(self, name: str, parameters: dict):
#         if not self.section_exists("connections"):
#             self._initialize_connection_section()
#         self._find_section("connections").add(name, parameters)
#         self._dump_config()
#
#     def get_logs_config(self) -> dict:
#         logs_config = _DEFAULT_LOGS_CONFIG.copy()
#         if self.section_exists("logs"):
#             logs_config.update(**self.get_section("logs"))
#         return logs_config
#
#     def is_default_logs_path(self, path: Path) -> bool:
#         return path.resolve() == Path(str(_DEFAULT_LOGS_CONFIG["path"])).resolve()
#
#     def _add_options(self):
#         self.add_option(
#             name="options",
#             parse_str=tomlkit.parse,
#         )
#         self.add_option(
#             name="connections",
#             parse_str=tomlkit.parse,
#         )
#         self.add_option(
#             name="snowcli",
#             parse_str=tomlkit.parse,
#         )
#         self.add_option(
#             name="logs",
#             parse_str=tomlkit.parse,
#         )
#
#     def _find_section(self, *path) -> TOMLDocument:
#         section = self
#         idx = 0
#         while idx < len(path):
#             section = section[path[idx]]
#             idx += 1
#         return section
#
#     def _merge_section_with_env(
#         self, section: Union[Table, Any], *path
#     ) -> Dict[str, str]:
#         if isinstance(section, Table):
#             env_variables = self._get_envs_for_path(*path)
#             section_copy = section.copy()
#             section_copy.update(env_variables)
#             return section_copy.unwrap()
#         # It's a atomic value
#         return section
#
#     def _get_env_value(self, *path, key: str):
#         env_variable_name = (
#             "SNOWFLAKE_" + "_".join(p.upper() for p in path) + f"_{key.upper()}"
#         )
#         return os.environ.get(env_variable_name)
#
#     def _get_envs_for_path(self, *path) -> dict:
#         env_variables_prefix = "SNOWFLAKE_" + "_".join(p.upper() for p in path)
#         return {
#             k.replace(f"{env_variables_prefix}_", "").lower(): os.environ[k]
#             for k in os.environ.keys()
#             if k.startswith(env_variables_prefix)
#         }
#
#     def _initialize_connection_section(self):
#         self.conf_file_cache.append("connections", table())
#
#     def _initialize_logs_section(self):
#         logs_table = table()
#         logs_table.update(_DEFAULT_LOGS_CONFIG)
#         self.conf_file_cache.append("logs", logs_table)
#
#     def _initialise_config(self):
#         os.makedirs(os.path.dirname(self.file_path), exist_ok=True)
#         self.conf_file_cache = TOMLDocument()
#         self._initialize_connection_section()
#         self._initialize_logs_section()
#         self._dump_config()
#         log.info(f"Created Snowflake configuration file at {cli_config.file_path}")
#
#     def _dump_config(self):
#         with open(self.file_path, "w+") as fh:
#             dump(self.conf_file_cache, fh)
# =======
CONFIG_MANAGER.add_option(
    name="snowcli",
    parse_str=tomlkit.parse,
    default=dict(),
)


def config_init(config_file: Path):
    """
    Initializes the app configuration. Config provided via cli flag takes precedence.
    If config file does not exist we create an empty one.
    """
    if not config_file:
        return

    CONFIG_MANAGER.file_path = config_file
    if not config_file.exists():
        _initialise_config(config_file)
    CONFIG_MANAGER.read_config()
    print(get_logs_config())


def add_connection(name: str, parameters: dict):
    conf_file_cache = CONFIG_MANAGER.conf_file_cache
    if conf_file_cache.get("connections") is None:
        conf_file_cache["connections"] = {}
    conf_file_cache["connections"][name] = parameters
    _dump_config(conf_file_cache)


def _initialise_logs_section():
    conf_file_cache = CONFIG_MANAGER.conf_file_cache
    if conf_file_cache.get("logs") is None:
        conf_file_cache["logs"] = _DEFAULT_LOGS_CONFIG
    _dump_config(conf_file_cache)


_DEFAULT_LOGS_CONFIG = {
    "save_logs": False,
    "path": str(CONFIG_MANAGER.file_path.parent / "logs"),
    "level": "info",
}


def get_logs_config() -> dict:
    print(CONFIG_MANAGER.file_path)
    logs_config = _DEFAULT_LOGS_CONFIG.copy()
    if config_section_exists("snowcli", "logs"):
        logs_config.update(**get_config_section("snowcli", "logs"))
    return logs_config


def is_default_logs_path(path: Path) -> bool:
    return path.resolve() == Path(str(_DEFAULT_LOGS_CONFIG["path"])).resolve()


def connection_exists(connection_name: str) -> bool:
    return config_section_exists("connections", connection_name)


def config_section_exists(*path) -> bool:
    try:
        _find_section(*path)
        return True
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        return False


def get_connection(connection_name: str) -> dict:
    try:
        return get_config_section("connections", connection_name)
    except KeyError:
        raise MissingConfiguration(f"Connection {connection_name} is not configured")


def get_config_section(*path) -> dict:
    section = _find_section(*path)
    if type(section) is Container:
        return {s: _merge_section_with_env(section[s], *path, s) for s in section}
    elif type(section) is Table:
        return _merge_section_with_env(section, *path)
    raise UnsupportedConfigSectionTypeError(type(section))


def get_config_value(*path, key: str, default: Optional[Any] = Empty) -> Any:
    """Looks for given key under nested path in toml file."""
    env_variable = _get_env_value(*path, key=key)
    if env_variable:
        return env_variable
    try:
        return get_config_section(*path)[key]
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        if default is not Empty:
            return default
        raise


def _initialise_config(config_file: Path) -> None:
    os.makedirs(os.path.dirname(config_file), exist_ok=True)
    config_file.touch()
    _initialise_logs_section()
    log.info(f"Created Snowflake configuration file at {CONFIG_MANAGER.file_path}")


def _get_env_value(*path, key: str) -> str | None:
    env_variable_name = (
        "SNOWFLAKE_" + "_".join(p.upper() for p in path) + f"_{key.upper()}"
    )
    return os.environ.get(env_variable_name)


def _find_section(*path) -> TOMLDocument:
    section = CONFIG_MANAGER
    idx = 0
    while idx < len(path):
        section = section[path[idx]]
        idx += 1
    return section


def _merge_section_with_env(section: Union[Table, Any], *path) -> Dict[str, str]:
    if isinstance(section, Table):
        env_variables = _get_envs_for_path(*path)
        section_copy = section.copy()
        section_copy.update(env_variables)
        return section_copy.unwrap()
    # It's a atomic value
    return section


def _get_envs_for_path(*path) -> dict:
    env_variables_prefix = "SNOWFLAKE_" + "_".join(p.upper() for p in path)
    return {
        k.replace(f"{env_variables_prefix}_", "").lower(): os.environ[k]
        for k in os.environ.keys()
        if k.startswith(env_variables_prefix)
    }


def _dump_config(conf_file_cache: Dict):
    with open(CONFIG_MANAGER.file_path, "w+") as fh:
        dump(conf_file_cache, fh)
