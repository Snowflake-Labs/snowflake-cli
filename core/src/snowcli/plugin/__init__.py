from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

from snowcli.plugin.api import PluginCommandGroupSpec


class PluginLoadingMode(Enum):
    ALL_INSTALLED_PLUGINS = 1
    ONLY_ENABLED_PLUGINS = 2


@dataclass
class PluginPackageInfo:
    package_name: str
    package_version: str
    dependencies: List[str]


@dataclass
class LoadedPlugin:
    plugin_name: str
    is_enabled: bool
    plugin_package_info: PluginPackageInfo
    command_group_spec: PluginCommandGroupSpec


@dataclass
class PluginAfterVerification:
    plugin: LoadedPlugin
    verification_error: Optional[Exception]

    def is_ok(self) -> bool:
        return self.verification_error is None

    def error_str(self) -> Optional[str]:
        error = self.verification_error
        if error:
            return f"{error.__class__.__name__} - {error.__str__()}"
        else:
            return None
