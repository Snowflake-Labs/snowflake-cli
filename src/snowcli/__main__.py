from __future__ import annotations

import sys

from snowcli.cli.app import app


if __name__ == "__main__":
    app()

if getattr(sys, "frozen", False):
    app(sys.argv[1:])
