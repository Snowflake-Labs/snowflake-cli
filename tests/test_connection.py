import os
from tempfile import NamedTemporaryFile
from textwrap import dedent
from unittest import mock

from snowcli.config import CliConfigManager


def test_new_connection_can_be_added(runner, snapshot):
    with NamedTemporaryFile("w+", suffix=".toml") as tmp_file:
        result = runner.invoke(
            [
                "--config-file",
                tmp_file.name,
                "connection",
                "add",
                "--connection",
                "conn1",
                "--username",
                "user1",
                "--password",
                "password1",
                "--account",
                "account1",
            ]
        )
        content = tmp_file.read()
    assert result.exit_code == 0, result.output
    assert content == snapshot


def test_fails_if_existing_connection(runner):
    with NamedTemporaryFile("w+", suffix=".toml") as tmp_file:
        tmp_file.write(
            dedent(
                """\
        [connections]
        [connections.conn2]
        username = "foo"
        """
            )
        )
        tmp_file.flush()
        result = runner.invoke(
            [
                "--config-file",
                tmp_file.name,
                "connection",
                "add",
                "--connection",
                "conn2",
                "--username",
                "user1",
                "--password",
                "password1",
                "--account",
                "account1",
            ]
        )
    assert result.exit_code == 1
    assert "Connection conn2 already exists  " in result.output


def test_environment_variables_override_configuration_value(test_snowcli_config):
    cm = CliConfigManager(file_path=test_snowcli_config)
    cm.read_config()

    assert cm.get("connections", "default", key="warehouse") == "xs"
    with mock.patch.dict(
        os.environ, {"SNOWFLAKE_CONNECTIONS_DEFAULT_WAREHOUSE": "foo42"}
    ):
        assert cm.get("connections", "default", key="warehouse") == "foo42"
