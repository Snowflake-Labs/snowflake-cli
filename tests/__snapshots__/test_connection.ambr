# serializer version: 1
# name: test_connection_can_be_added_with_existing_paths_in_arguments[-k]
  '''
  Snowflake password [optional]: 
  Role for the connection [optional]: 
  Warehouse for the connection [optional]: 
  Database for the connection [optional]: 
  Schema for the connection [optional]: 
  Connection host [optional]: 
  Snowflake region [optional]: 
  Authentication method [optional]: 
  Path to token file [optional]: 
  Wrote new connection conn1 to /Users/jsikorski/.snowflake/config.toml
  
  '''
# ---
# name: test_connection_can_be_added_with_existing_paths_in_arguments[-t]
  '''
  Snowflake password [optional]: 
  Role for the connection [optional]: 
  Warehouse for the connection [optional]: 
  Database for the connection [optional]: 
  Schema for the connection [optional]: 
  Connection host [optional]: 
  Snowflake region [optional]: 
  Authentication method [optional]: 
  Path to private key file [optional]: 
  Wrote new connection conn1 to /Users/jsikorski/.snowflake/config.toml
  
  '''
# ---
# name: test_connection_can_be_added_with_existing_paths_in_prompt[10]
  '''
  [connections.connName]
  account = "accName"
  user = "userName"
  password = "password"
  token_file_path = "/var/folders/k8/3sdqh3nn4gg7lpr5fz0fjlqw0000gn/T/tmpjbd8o_i2"
  
  '''
# ---
# name: test_connection_can_be_added_with_existing_paths_in_prompt[9]
  '''
  [connections.connName]
  account = "accName"
  user = "userName"
  password = "password"
  private_key_path = "/var/folders/k8/3sdqh3nn4gg7lpr5fz0fjlqw0000gn/T/tmp0rnw_ay8"
  
  '''
# ---
# name: test_file_paths_have_to_exist_when_given_in_arguments[-k]
  '''
  +- Error ----------------------------------------------------------------------+
  | Path ~/path/to/file does not exist.                                          |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_file_paths_have_to_exist_when_given_in_arguments[-t]
  '''
  +- Error ----------------------------------------------------------------------+
  | Path ~/path/to/file does not exist.                                          |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_file_paths_have_to_exist_when_given_in_prompt[10]
  ''
# ---
# name: test_file_paths_have_to_exist_when_given_in_prompt[9]
  ''
# ---
# name: test_if_whitespaces_are_stripped_from_connection_name
  '''
  [connections.whitespaceTest]
  account = "accName"
  user = "userName"
  password = "123"
  host = "baz"
  region = "Kaszuby"
  port = "12345"
  database = "foo"
  schema = "bar"
  warehouse = "some warehouse"
  role = "some role"
  authenticator = " foo   "
  private_key_path = ""
  
  '''
# ---
# name: test_if_whitespaces_are_stripped_from_connection_name.1
  '''
  [
      {
          "connection_name": "whitespaceTest",
          "parameters": {
              "account": "accName",
              "user": "userName",
              "password": "****",
              "host": "baz",
              "region": "Kaszuby",
              "port": "12345",
              "database": "foo",
              "schema": "bar",
              "warehouse": "some warehouse",
              "role": "some role",
              "authenticator": " foo   ",
              "private_key_path": ""
          },
          "is_default": false
      }
  ]
  
  '''
# ---
# name: test_new_connection_add_prompt_handles_default_values
  '''
  [connections.connName]
  account = "accName"
  user = "userName"
  
  '''
# ---
# name: test_new_connection_add_prompt_handles_prompt_override
  '''
  [connections.connName]
  account = "accName"
  user = "userName"
  password = "dbName"
  
  '''
# ---
# name: test_new_connection_can_be_added
  '''
  [connections.conn1]
  account = "account1"
  user = "user1"
  password = "password1"
  port = "8080"
  
  '''
# ---
# name: test_new_connection_can_be_added_as_default
  '''
  default_connection_name = "default-conn"
  [connections.default-conn]
  account = "account1"
  user = "user1"
  password = "password1"
  
  '''
# ---
# name: test_new_connection_with_jwt_auth
  '''
  [connections.conn2]
  account = "account1"
  user = "user2"
  authenticator = "SNOWFLAKE_JWT"
  private_key_path = "~/private_key"
  
  '''
# ---
# name: test_second_connection_not_update_default_connection
  '''
  [connections]
  [connections.conn]
  username = "foo"
  
  [connections.conn2]
  account = "account1"
  user = "user1"
  password = "password1"
  [options]
  default_connection = "conn"
  
  '''
# ---
